### PR TITLE
Implementing Atomic features via DBUS interface

### PR DESCRIPTION
**Pull request by AmartC**
Implemented Atomic features using DBUS on atomic dbus server.

### CODE REVIEW

File: Atomic/diffDbus.py [hunk 1]
```diff
@@ -0,0 +1,406 @@
+import os
```
Comment (rhatdan) [inline Atomic/diffDbus.py]: THis should not be a new file.  Your changes should be merged into diff.py.

File: atomic_client.py [hunk 2]
```diff
@@ -15,21 +15,143 @@ def __init__(self):
     @polkit.enable_proxy
     def version(self, image, recurse):
         ret = self.dbus_object.version(image, recurse,
-                                       dbus_interface="org.atomic")
+                                        dbus_interface="org.atomic")
         return ret
```
Comment (rhatdan) [inline atomic_client.py]: Cleanup white space before submitting changes.

File: atomic_client.py [hunk 3]
```diff
@@ -15,21 +15,143 @@ def __init__(self):
...
     try:
         dbus_proxy = AtomicDBus()
-        resp = dbus_proxy.version(sys.argv[1:], True)
+    except dbus.DBusException as e:
+        print e
+
+    if(sys.argv[1] == "version"):
+        if sys.argv[2] == "-r":
+            try:
+                resp = dbus_proxy.version(sys.argv[3:], True)
```
Comment (rhatdan) [inline atomic_client.py]: Try block should wrap all functions.

File: atomic_dbus.py [hunk 4]
```diff
@@ -6,20 +6,42 @@
...
+            self.rootfs = None
+            self.json = False
+            self.no_files = False
+            self.names_only = False
+            self.rpms = False
+            self.verbose = False
+            self.containers = None
+            self.optional = None
+            self.d = None
+            self.n = None
```
Comment (rhatdan) [inline atomic_dbus.py]: Do we need all of these defined in init, or can we just define them for each interface that needs them

File: atomic_dbus.py [hunk 5]
```diff
@@ -44,16 +67,125 @@ def version(self, images, recurse=False):
...
+        # No arguments are passed for storage_reset function
+        args = self.Args()
+        storage.set_args(args)
+        storage.reset()
+
+    """
+    The storage_import method imports all containers and their associated contents from a filesystem directory.
+    """
+    @slip.dbus.polkit.require_auth("org.atomic.read")
+    @dbus.service.method("org.atomic", in_signature='ss', out_signature='')
```
Comment (rhatdan) [inline atomic_dbus.py]: Need to change the polkit to org.atomic.write

File: atomic_dbus.py [hunk 6]
```diff
@@ -44,16 +67,125 @@ def version(self, images, recurse=False):
...
+    @dbus.service.method("org.atomic", in_signature='asasbbasbasb',
+                         out_signature='')
+    def scan(self, scan_targets=[], scanners=None, _list=False, verbose=False, rootfs=[], all=False, images=False, containers=False ):
+        scan = Scan()
+        args = self.Args()
+        args.scan_targets = scan_targets
+        args.scanners = scanners
+        args.list = _list
+        args.verbose = verbose
+        args.rootfs - rootfs
```
Comment (rhatdan) [inline atomic_dbus.py]: What does args.rootfs - rootfs do?

File: atomic_dbus.py [hunk 7]
```diff
@@ -44,16 +67,125 @@ def version(self, images, recurse=False):
...
+    """
+    @slip.dbus.polkit.require_auth("org.atomic.read")
+    @dbus.service.method("org.atomic", in_signature='asasbbasbasb',
+                         out_signature='')
+    def scan(self, scan_targets=[], scanners=None, _list=False, verbose=False, rootfs=[], all=False, images=False, containers=False ):
+        scan = Scan()
+        args = self.Args()
+        args.scan_targets = scan_targets
+        args.scanners = scanners
+        args.list = _list
```
Comment (rhatdan) [inline atomic_dbus.py]: Change _list to _list_scanners, but this should be a separate function.  scan_list().  Not used for scanning.

File: atomic_dbus.py (hunk 6 above)
Comment (AmartC) [inline atomic_dbus.py]: My mistake. should have been args.rootfs = rootfs. I fixed that.

File: Atomic/diff.py [hunk 8]
```diff
@@ -2,24 +2,63 @@
 import sys
 import rpm
 from filecmp import dircmp
-
 from . import util
```
Comment (rhatdan) [inline Atomic/diff.py]: Why did you delete this line?

File: Atomic/diff.py [hunk 9]
```diff
@@ -2,24 +2,63 @@
...
 from . import mount
 from . import Atomic
 
+''' This file will handle both dbus calls and regular calls from Atomic '''
+
```
Comment (rhatdan) [inline Atomic/diff.py]: No reason for this, it has no idea that dbus is calling it.  Just remove comment.

File: Atomic/diff.py [hunk 10]
```diff
@@ -2,24 +2,63 @@
...
 class Diff(Atomic):
+
+    def diff_tty(self):
+        '''
+        Prints diff information depending on what parameters are passed by the user.
+        :return: None
```
Comment (rhatdan) [inline Atomic/diff.py]: Prints diff information to the tty ...

File: Atomic/diff.py [hunk 11]
```diff
@@ -33,48 +72,143 @@ def diff(self):
...
-                util.output_json(helpers.json_out)
+            return diff_dict
+
         except KeyboardInterrupt:
             util.write_out("Quitting...")
             helpers._cleanup(image_list)
 
+
+    def output_files(self, images, diff_dict):
+        '''
```
Comment (rhatdan) [inline Atomic/diff.py]: Could you put output_files back to where it was originally so we get a better idea of the diffs.  Moving these methods around causes review to be too difficult.

File: atomic_dbus.py [hunk 12]
```diff
@@ -47,13 +69,114 @@ def version(self, images, recurse=False):
...
+        storage.modify()
+
+    """
+    The diff method shows differences between two container images, file diff or RPMS.
+    """
+    @slip.dbus.polkit.require_auth("org.atomic.read")
+    @dbus.service.method("org.atomic", in_signature='ssb',
+                         out_signature='a{sa{sas}}')
+    def diff(self,first,second, rpms=False):
+        diff = Diff()
```
Comment (rhatdan) [inline atomic_dbus.py]: Space separated list

File: atomic_dbus.py [hunk 13]
```diff
@@ -47,13 +69,114 @@ def version(self, images, recurse=False):
...
+        args = self.Args()
+        args.compares.append(first)
+        args.compares.append(second)
+        args.json = False
```
Comment (rhatdan) [inline atomic_dbus.py]: args.compares = [first, second]

**Comment by rhatdan** [comment]
You need to squash your pull requests and probably give a different one for each command in the dbus interface.

One patch for diff, scan, scan-list ...
